inference_perf/client/modelserver/openai_client.py

```diff
@@ -0,0 +1,157 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import abstractmethod
+from inference_perf.client.requestdatacollector import RequestDataCollector
+from inference_perf.config import APIConfig, APIType, CustomTokenizerConfig
+from inference_perf.apis import InferenceAPIData, InferenceInfo, RequestLifecycleMetric, ErrorResponseInfo
+from inference_perf.utils import CustomTokenizer
+from .base import ModelServerClient, PrometheusMetricMetadata
+from typing import List, Optional
+import aiohttp
+import json
+import time
+import logging
+import requests
+
+logger = logging.getLogger(__name__)
+
+
+class openAIModelServerClient(ModelServerClient):
+    def __init__(
+        self,
+        metrics_collector: RequestDataCollector,
+        api_config: APIConfig,
+        uri: str,
+        model_name: Optional[str],
+        tokenizer_config: Optional[CustomTokenizerConfig],
+        max_tcp_connections: int,
+        additional_filters: List[str],
+        ignore_eos: bool = True,
+        api_key: Optional[str] = None,
+    ) -> None:
+        super().__init__(api_config)
+        self.uri = uri
+        self.max_completion_tokens = 30  # default to use when not set at the request level
+        self.ignore_eos = ignore_eos
+        self.metrics_collector = metrics_collector
+        self.max_tcp_connections = max_tcp_connections
+        self.additional_filters = additional_filters
+        self.api_key = api_key
+
+        if model_name is None:
+            supported_models = self.get_supported_models()
+            if not supported_models:
+                logger.error("No supported models found")
+                raise Exception("openAI client init failed, no model_name could be found")
+            self.model_name = supported_models[0].get("id")
+            logger.info(f"Inferred model {self.model_name}")
+            if len(supported_models) > 1:
+                logger.warning(f"More than one supported model found {supported_models}, selecting {self.model_name}")
+        else:
+            self.model_name = model_name
+
+        if tokenizer_config and not tokenizer_config.pretrained_model_name_or_path:
+            tokenizer_config.pretrained_model_name_or_path = self.model_name
+        elif not tokenizer_config:
+            tokenizer_config = CustomTokenizerConfig(pretrained_model_name_or_path=self.model_name)
+        self.tokenizer = CustomTokenizer(tokenizer_config)
+
+        self.additional_metric_filters = [f"model_name='{self.model_name}'", *self.additional_filters]
+
+    async def process_request(self, data: InferenceAPIData, stage_id: int, scheduled_time: float) -> None:
+        payload = data.to_payload(
+            model_name=self.model_name,
+            max_tokens=self.max_completion_tokens,
+            ignore_eos=self.ignore_eos,
+            streaming=self.api_config.streaming,
+        )
+        headers = {"Content-Type": "application/json"}
```

Comment:
Looks like this is missing my recently merged custom header change. Would be good to rebase and include it.

Reply:
Rebased and updated in [d168677](https://github.com/kubernetes-sigs/inference-perf/pull/193/commits/d1686777fd8d938d9ba255c4141e2f8e80069e13)